numpy/core/fromnumeric.py

```diff
@@ -3841,14 +3855,18 @@ def product(*args, **kwargs):
     --------
     prod : equivalent function; see for details.
     """
-    # 2023-03-02, 1.25.0
-    warnings.warn("`product` is deprecated as of NumPy 1.25.0, and will be "
-                  "removed in NumPy 2.0. Please use `prod` instead.",
-                  DeprecationWarning, stacklevel=2)
     return prod(*args, **kwargs)
```

Comment:
I think I hinted at the solution (and stacking uses it), it is something like:
```
    if not overrides.ARRAY_FUNCTION_ENABLED:
        # call dispatcher explicitly, as it gives a deprecation warning
        _product_dispatcher(a, axis=axis, dtype=dtype, out=out)
```

Reply:
Ah yes, thanks. 

Adding deprecations like this is pretty cumbersome - but having the pattern now helps. It did raise the question for me whether we want to keep the feature of disabling `__array_function__` dispatching permanently. If so, we should change `NUMPY_EXPERIMENTAL_ARRAY_FUNCTION` to something more suitable I think. And if not, we can remove it in the 2.0 release perhaps. WDYT?

Reply:
Yeah, I was wondering that too...  I wouldn't even care to remove it in a minor realse, it is seems like a pretty niche feature and easy to add back if someone shouts.

There may be niche use-cases where people actually do disable it to remove overhead, and I suppose and in some rare cases it may even help.  But overall, it has a discoverability of about zero, so I don't really see more than 5 users in the world...

Reply:
I'll follow up on this one on the mailing list then.